EIPS/eip-7971.md

```diff
@@ -0,0 +1,159 @@
+
+### Transaction-Global Transient Storage Limit
+
+A transaction-global counter tracks the number of unique transient storage slots written across all contracts during transaction execution:
+
+1. At the beginning of each transaction, initialize a counter `transient_slots_used` to 0.
+2. When `TSTORE` is executed:
+   - If the slot has not been written to during this transaction (across any contract), increment `transient_slots_used`.
+   - If `transient_slots_used` exceeds `MAX_TRANSIENT_SLOTS`, the transaction MUST exceptionally halt.
+3. The counter persists across all message calls in the transaction.
+4. The counter is reset to 0 at the end of the transaction.
+
+### Implementation Note
+
+Implementations MUST track transient storage allocated across all contracts. A slot is considered unique based on the tuple `(contract_address, storage_key)`. Writing to the same slot multiple times within a transaction does not generally increment the counter after the first write (unless the slot gets deallocated with a revert).
+
+## Rationale
+
+### Constant Pricing with Hard Limit
+
+This EIP implements constant pricing with a hard limit for several reasons:
+
+1. A hard limit provides guarantees on the total resource consumption which are easier to reason about for clients, rather than needing to perform a calculation as a function of current gas limits to find out what total memory consumption could be.
+2. Common use cases (e.g., reentrancy locks using 1-2 slots) are not penalized for worst case resource usage (like in DOS attacks).
+3. Clients can safely pre-reserve all memory which could be used by transient storage up-front at the beginning of a transaction.
+
+### Gas Cost Selection
+
+- `GAS_TLOAD` (5 gas): Transient storage reads require only memory access without disk I/O.
+- `GAS_WARM_SLOAD` (5 gas): Warm storage loads from cache have similar performance characteristics to transient storage reads.
+- `GAS_TSTORE` (12 gas): Transient storage writes require memory allocation and journaling for revert support.
+- `GAS_TSTORE_ALLOCATE` (24 gas): Writes to fresh slots require memory allocation, which is more expensive than writing to an existing slot.
+
+### Hard Limit Selection
+
+`MAX_TRANSIENT_SLOTS` of 131072 allows:
+
+- Sufficient slots for typical user applications
+- Memory usage bounded to approximately 8 MB per transaction (131072 slots * 64 bytes), which prevents OOM-based denial-of-service attacks
+
+### Design Alternatives Considered
+
+1. Per-Contract Limits: Increased complexity in reasoning about resource consumption based on the shape of the call stack.
+2. Superlinear Pricing: Adds complexity, and still punishes "common" (non-DOS) use cases.
+3. No Limit: May allow memory-based DOS attack if transaction-level gas limits change, or if pricing changes in the future.
+
+The benefit of a hard limit is that the resource consumption is bounded in a predictable way even in the presence of other parameter changes in the protocol.
```

Comment:
I agree with the motivation to make this forward-compatible so there is protection if we start tuning the gas parameters again. However, with https://eips.ethereum.org/EIPS/eip-7825 we introduce the 2^24 gas limit which means we can create 2^24 // 36 = 466033 transient storage slots. This is slightly less than 4 times the current cap and would allocate 29_826_112 bytes of memory at most.

I'm not sure, 8 MB sounds like a reasonable limit, 30 might be too high if you start to execute txs in parallel (e.g. from mempool or RPC calls). Will leave a final comment when I'm done reviewing on this point.